core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -496,14 +496,16 @@ public void updateShardState(final ShardRouting newRouting,
                                     public void onResponse(ResyncTask resyncTask) {
                                         logger.info("primary-replica resync completed with {} operations",
                                             resyncTask.getResyncedOperations());
-                                        boolean resyncCompleted = primaryReplicaResyncInProgress.compareAndSet(true, false);
-                                        assert resyncCompleted : "primary-replica resync finished but was not started";
+                                        synchronized (mutex) {
+                                            final IndexShardState prevState = changeState(IndexShardState.STARTED, "Resync is completed");
+                                            if (prevState != IndexShardState.PROMOTING) {
+                                                throw new IllegalIndexShardStateException(shardId, prevState, "primary-replica resync finished but was not started");
+                                            }
+                                        }
                                     }
 
                                     @Override
                                     public void onFailure(Exception e) {
-                                        boolean resyncCompleted = primaryReplicaResyncInProgress.compareAndSet(true, false);
```

Comment:
as we have no corresponding state transition here now, does it mean that the shard can forever be stuck in PROMOTING state, allow no relocations?